src/dtype.rs

```diff
@@ -57,6 +58,22 @@ pub fn dtype<T: Element>(py: Python) -> &PyArrayDescr {
 }
 
 impl PyArrayDescr {
+    /// Creates a new dtype object from an arbitrary object.
+    ///
+    /// Equivalent to invoking the constructor of `np.dtype`.
+    pub fn new<'py>(py: Python<'py>, obj: &(impl ToPyObject + ?Sized)) -> PyResult<&'py Self> {
+        let obj = obj.to_object(py);
+        let mut descr: *mut PyArray_Descr = ptr::null_mut();
+        unsafe {
+            // None is an invalid input here and is not converted to NPY_DEFAULT_TYPE
+            PY_ARRAY_API.PyArray_DescrConverter2(obj.as_ptr(), &mut descr as *mut _);
+        }
+        if descr.is_null() {
+            return Err(PyErr::fetch(py));
+        }
+        Ok(unsafe { py.from_owned_ptr(descr as _) })
```

Comment:
I think this could be simplified using [`Python::from_owned_ptr_or_err`](https://docs.rs/pyo3/0.15.1/pyo3/prelude/struct.Python.html#method.from_owned_ptr_or_err).

Reply:
Good call, indeed